http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java

```diff
@@ -106,14 +110,52 @@ protected void channelRead0(ChannelHandlerContext channelContext, HttpObject msg
 
                 fullContent = ((FullHttpResponse) msg).content();
                 ByteBuffer bb = copyToByteBuffer(fullContent);
-                requestContext.handler().onStream(new FullResponseContentPublisher(channelContext, bb, ef));
-                finalizeResponse(requestContext, channelContext);
+                requestContext.handler().onStream(new DataCountingPublisher(channelContext,
+                                                                            new FullResponseContentPublisher(channelContext,
+                                                                                                             bb, ef)));
+
+                try {
+                    validateResponseContentLength(channelContext);
+                    finalizeResponse(requestContext, channelContext);
```

Comment:
If `validateResponseContentLength` throws we do not call `finalizeResponse`. Likewise for streaming. Do we still release the channel in that scenario?

Reply:
Yep, exceptionCaught closes and releases the channel.